build/scripts/script.js

```diff
@@ -591,6 +579,113 @@ function updateSearchResults(value) {
   }
 }
 
+/**
+ * Add missing top sections headers (h1) in search results.
+ * Search result can be h1, h2 or h3, but the search algorithm can
+ * find a match for a h3 section and not for the h1.
+ * To not display h3 orphan in the tree view, we manually to
+ * the search result the h1 that is a parent to that h3.
+ * @param {array} searchResults
+ * @returns
+ */
+function addMissingTopSections(searchResults) {
+  const h1Sections = {};
+  const missingH1 = [];
+  for (result of searchResults) {
+    if (!h1Sections[result.doc.h1]) {
+      h1Sections[result.doc.h1] = [];
+    }
+    h1Sections[result.doc.h1].push(result);
+  }
+
+  for (const value of Object.values(h1Sections)) {
+    const topSection = value.find((r) => {
+      return r.doc.h2 === undefined;
+    });
+    if (topSection === undefined) {
+      const fakeH1 = structuredClone(value[0]);
+      fakeH1.doc.h2 = undefined;
+      fakeH1.doc.h3 = undefined;
+      fakeH1.doc.body = "";
+      missingH1.push(fakeH1);
+    }
+  }
+  return searchResults.concat(missingH1);
+}
+
+/**
+ * Re-orders search results by grouping them according
+ * to their shared section headings (h1, h2, h3).
+ * @param {array} searchResults The array of search results to be re-ordered.
+ * @returns An array re-ordered based on section headings.
+ *
+ * @example
+ */
+function reorderSearchResultByGroup(searchResults) {
+  const results = addMissingTopSections(searchResults);
+
+  // group by h1
+  const groupedSearchResult = groupArrayItemsBy(results, (item) => item.doc.h1);
```

Comment:
Here also, don't we risk grouping multiple h1 with the same name but in different pages?